File: proj 1/guitar.py
class Guitar:
    DEFAULT_STRING_COUNT : int = 6
    DEFAULT_STRING_GAUGE : int = 3221 
    DEFAULT_TUNING : str = "E A D G B E"
    ACCEPTABLE_NOTES : list = ['Ab','A','A#','Bb','B','C','C#','Db','D','D#','Eb','E','F','F#','Gb','G','G#']
    DEFAULT_VOLUME : float = 0.0

    def __init__(self, tuning : str = DEFAULT_TUNING, stringCount : int = DEFAULT_STRING_COUNT, stringGauge : int = DEFAULT_STRING_GAUGE, volume : float = DEFAULT_VOLUME):
      self._stringCount = stringCount
      self._stringGauge = stringGauge
      self._tuning = tuning
      self._volume = volume

    @property
    def stringCount(self):
        return self._stringCount

    @stringCount.setter
    def stringCount(self, numStrings : int):
        if numStrings < 0:
            numStrings = 0
        self._stringCount = numStrings

    @property
    def stringGauge(self):
        return self._stringGauge
    
    @stringGauge.setter
    def stringGauge(self, strGauge : int):
        self._stringGauge = strGauge

    @property
    def tuning(self):
        return self._tuning

    @tuning.setter
    def tuning(self, vals : str):
        newVals = list(vals.split(" "))
        if len(newVals) != self._stringCount:
            vals = self.DEFAULT_TUNING
        for note in newVals:
            if note not in self.ACCEPTABLE_NOTES:
                vals = self.DEFAULT_TUNING
        self._tuning = vals


    @property
    def volume(self):
        return self._volume

    @volume.setter
    def volume(self, vol : float):
        if (vol < 0):
            vol = 0
        elif (vol > 10):
            vol = 10
        self._volume = vol

File: proj_1_cpp/src/Guitar.cpp
#include <stdexcept>
#include "Guitar.h"
#include <iostream>
namespace instrument {
    const int Guitar::DEFAULT_STRING_COUNT = 6;
    const int Guitar::DEFAULT_STRING_GAUGE = 3221;
    const double Guitar::DEFAULT_VOLUME = 0.0;
    const std::string Guitar::DEFAULT_TUNING = "E A D G B E";
    const char Guitar::ACCEPTABLE_NOTES[7] = {'A','B','C','D','E','F','G'};

    Guitar::Guitar(int _stringCount, int _stringGauge, std::string _tuning, double _volume)
        : stringCount(_stringCount), stringGauge(_stringGauge), tuning(_tuning), volume(_volume) 
    {
        bool acceptTuning = true;
        bool notesIn = false;
        char tempStr = ' ';
        int countStr = 0;
        if(stringCount < 0) stringCount = 0;
        if (volume < 0) {
            volume = 0.0;
        }        
        else if(volume > 10) {
            volume = 10.0;
        }
        for(int i = 0; i < tuning.length(); i++) {
            if(tuning[i] != ' ') {
                //the following two switch statements take and sharp or flat notes and check to make sure
                //they are real notes. Since C flat is not a real note, I change it to B, since E sharp
                //is not a real note I change it to F
                //I also decrease the counter on the string so as to not double count any string
                if(tuning[i] == 'b') {
                    countStr--;
                    switch (tuning[i - 1])
                    {
                    case 'C':
                        tempStr = 'B';
                        break;
                    case 'F':
                        tempStr = 'E';
                        break;
                    default:
                        tempStr = tuning[i - 1];
                        break;
                    }
                }
                else if(tuning[i] == '#') {
                    countStr--;
                    switch (tuning[i - 1])
                    {
                        case 'B':
                            tempStr = 'C';
                            break;
                        case 'E':
                            tempStr = 'F';
                            break;
                        default:
                            tempStr = tuning[i - 1];
                            break;
                    }
                }
                else {
                        tempStr = tuning[i];
                }
                for(int j = 0; j < 7; j++) {
                    if(tempStr == ACCEPTABLE_NOTES[j]) notesIn = true;
                }
                if(!notesIn) acceptTuning = false;
                else {
                    notesIn = false;
                    tempStr = ' ';
                    countStr++;
                }
            }
        }
        if(!acceptTuning || countStr != stringCount) {
            tuning = DEFAULT_TUNING;
        }
    }

    int Guitar::getStringCount() const {
        return stringCount;
    }

    void Guitar::setStringCount(int numStrings) {
        if(numStrings < 0) numStrings = 0;
        stringCount = numStrings;
    }

    int Guitar::getStringGauge() const {
        return stringGauge;
    }

    void Guitar::setStringGauge(int gauge) {
        stringGauge = gauge;
    }

    std::string Guitar::getTuning() const {
        return tuning;
    }

    void Guitar::setTuning(std::string newTuning) {
        bool acceptTuning = true;
        bool notesIn = false;
        char tempStr;
        int countStr = 0;
        for(int i = 0; i < newTuning.length(); i++) {
            if(newTuning[i] != ' ') {
                if(newTuning[i] == 'b') {
                    countStr--;
                    switch (newTuning[i - 1])
                    {
                    case 'C':
                        tempStr = 'B';
                        break;
                    case 'F':
                        tempStr = 'E';
                        break;
                    default:
                        tempStr = newTuning[i - 1];
                        break;
                    }
                }
                else if(newTuning[i] == '#') {
                    countStr--;
                    switch (newTuning[i - 1])
                    {
                        case 'B':
                            tempStr = 'C';
                            break;
                        case 'E':
                            tempStr = 'F';
                            break;
                        default:
                            tempStr = newTuning[i - 1];
                            break;
                    }
                }
                else {
                    tempStr = newTuning[i];
                }
                for(int j = 0; j < 7; j++) {
                    if(tempStr == ACCEPTABLE_NOTES[j]) notesIn = true;
                }
                if(!notesIn) acceptTuning = false;
                else {
                    notesIn = false;
                    tempStr = ' ';
                    countStr++;
                }
            }
        }
        if(!acceptTuning || countStr != stringCount) {
            newTuning = DEFAULT_TUNING;
        }
        tuning = newTuning;
    }

    double Guitar::getVolume() const {
        return volume;
    }

    void Guitar::setVolume(double newVol) {
        if (newVol < 0.0) {
            volume = 0.0;
        }
        else if(newVol > 10.0) {
            volume = 10.0;
        }
        else volume = newVol;
    }
}